sklearn/metrics/_classification.py

```diff
@@ -2979,32 +2981,37 @@ def log_loss(y_true, y_pred, *, normalize=True, sample_weight=None, labels=None)
                 "got {0}.".format(lb.classes_)
             )
 
-    transformed_labels = lb.transform(y_true)
+    float_dtype = _find_matching_floating_dtype(y_true, y_pred, xp=xp)
+    transformed_labels = xp.asarray(
+        lb.transform(_convert_to_numpy(y_true, xp=xp)),
+        dtype=float_dtype,
+        device=device_,
+    )
 
     if transformed_labels.shape[1] == 1:
-        transformed_labels = np.append(
-            1 - transformed_labels, transformed_labels, axis=1
+        transformed_labels = xp.concat(
+            (1 - transformed_labels, transformed_labels), axis=1
         )
 
     # If y_pred is of single dimension, assume y_true to be binary
     # and then check.
     if y_pred.ndim == 1:
-        y_pred = y_pred[:, np.newaxis]
+        y_pred = y_pred[:, xp.newaxis]
     if y_pred.shape[1] == 1:
-        y_pred = np.append(1 - y_pred, y_pred, axis=1)
+        y_pred = xp.concat((1 - y_pred, y_pred), axis=1)
 
-    eps = np.finfo(y_pred.dtype).eps
+    eps = xp.finfo(y_pred.dtype).eps
 
     # Make sure y_pred is normalized
-    y_pred_sum = y_pred.sum(axis=1)
-    if not np.allclose(y_pred_sum, 1, rtol=np.sqrt(eps)):
+    y_pred_sum = xp.sum(y_pred, axis=1)
+    if not _allclose(y_pred_sum, 1, rtol=np.sqrt(eps), xp=xp):
```

Comment:
**Note**: Since we are internally converting to numpy in all cases, we can use `np.sqrt` instead of `xp.sqrt` here as this is just a scalar value and using `xp.sqrt` will require us to unnecessarily convert `eps` into an array to satisfy array api strict.